test/e2e/clusterctl_upgrade.go

```diff
@@ -428,7 +430,30 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 			Name:                   workloadClusterName,
 		}, input.E2EConfig.GetIntervals(specName, "wait-cluster")...)
 
-		expectedMachineCount := *controlPlaneMachineCount + calculateExpectedWorkerCount(ctx, managementClusterProxy.GetClient(), workloadClusterUnstructured, coreCAPIStorageVersion)
+		var infraMachinePool *unstructured.Unstructured
+		skipMachinePoolMachines := false
+
+		By("Determining if MachinePool machines should be skipped")
+		machinePool := getMachinePool(ctx, managementClusterProxy.GetClient(), workloadClusterNamespace, workloadClusterName)
+		if machinePool != nil {
+			infraMachinePool = getInfraMachinePool(ctx, managementClusterProxy.GetClient(), *machinePool, workloadClusterNamespace, workloadClusterName)
+		}
+
+		if infraMachinePool != nil {
+			// Check if the InfraMachinePool has an infrastructureMachineKind field. If it does not, we should skip checking for MachinePool machines.
+			if err := util.UnstructuredUnmarshalField(infraMachinePool, ptr.To(""), "status", "infrastructureMachineKind"); err != nil {
```

Comment:
Is it safe to assume that at this point in time the InfraMachinePool controller already set this field if it's planning to do that? (I guess, yes)

Reply:
I think so, that's why we're checking if the field exists in order to know whether or not the MachinePool is backed by Machines.

Reply:
Yup, but can you differentiate between:
1) MachinePool doesn't support Machines
2) InfraMachinePool controller didn't reconcile yet, so it just didn't have a chance to set the status field yet


My guess is we're sort of safe it's not the second case because usually the MP controller is quicker